File: platform/common/src/main/kotlin/com/mdeo/common/transport/ExecutionWsClient.kt
package com.mdeo.common.transport

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.websocket.*
import io.ktor.websocket.*
import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.slf4j.LoggerFactory
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Request/response client for [ExecutionWsProtocol], over WebSocket connections that are
 * cached per target service and closed again once they go unused.
 *
 * A connection is opened on the first request to a target, shared by every request to that
 * target afterwards, and closed after [idleTimeoutMillis] without traffic. Nothing about a
 * request depends on connection state — each one carries its own authorization — so a
 * connection that was dropped in between is simply reopened, and two unrelated callers can
 * share one without either being able to observe the other.
 *
 * @param idleTimeoutMillis How long a connection may sit without traffic before it is closed
 * @param requestTimeoutMillis How long a single request may wait for its terminating response
 * @param httpClient WebSocket-capable client to dial with; a private one is created if omitted
 */
class ExecutionWsClient(
    private val idleTimeoutMillis: Long = DEFAULT_IDLE_TIMEOUT_MILLIS,
    private val requestTimeoutMillis: Long = DEFAULT_REQUEST_TIMEOUT_MILLIS,
    httpClient: HttpClient? = null
) : AutoCloseable {

    companion object {
        /**
         * Default idle lifetime of a pooled connection. Long enough that the bursts of
         * requests one user interaction produces all share a connection, short enough that
         * an idle service is not holding sockets open for sessions nobody is using.
         */
        const val DEFAULT_IDLE_TIMEOUT_MILLIS = 180_000L

        /**
         * Default per-request timeout. Bulk loads of a large result set are the slow case
         * this has to accommodate.
         */
        const val DEFAULT_REQUEST_TIMEOUT_MILLIS = 300_000L

        /**
         * How often the reaper looks for connections that have gone idle.
         */
        private const val REAPER_INTERVAL_MILLIS = 30_000L

        /**
         * Keepalive ping interval. A peer that vanishes without closing the connection would
         * otherwise leave this side holding a socket that will never answer.
         */
        private const val PING_INTERVAL_MILLIS = 30_000L

        /**
         * How long dialling a peer may take before it counts as unreachable.
         */
        private const val CONNECT_TIMEOUT_MILLIS = 10_000L

        /**
         * Converts an `http(s)` service base URL into the `ws(s)` URL of its execution endpoint.
         *
         * @param baseUrl Base URL of the target service
         * @return The WebSocket URL of that service's execution endpoint
         */
        fun toWebSocketUrl(baseUrl: String): String {
            val trimmed = baseUrl.trimEnd('/')
            val scheme = when {
                trimmed.startsWith("https://") -> "wss://" + trimmed.removePrefix("https://")
                trimmed.startsWith("http://") -> "ws://" + trimmed.removePrefix("http://")
                else -> trimmed
            }
            return scheme + ExecutionWsProtocol.ENDPOINT_PATH
        }
    }

    private val logger = LoggerFactory.getLogger(ExecutionWsClient::class.java)

    private val ownsHttpClient = httpClient == null
    private val client: HttpClient = httpClient ?: HttpClient(CIO) {
        install(WebSockets) {
            pingIntervalMillis = PING_INTERVAL_MILLIS
            extensions { installDeflate() }
        }
        engine {
            // Without this, dialling a peer that accepts the connection but never completes the
            // upgrade blocks the connect mutex indefinitely, and every request to every target
            // queues behind it with nothing logged. A stalled dial has to become a failure.
            endpoint {
                connectTimeout = CONNECT_TIMEOUT_MILLIS
                requestTimeout = CONNECT_TIMEOUT_MILLIS
            }
        }
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val connections = ConcurrentHashMap<String, PooledConnection>()
    private val connectMutex = Mutex()

    private val reaper = scope.launch {
        while (isActive) {
            delay(REAPER_INTERVAL_MILLIS)
            reapIdleConnections()
        }
    }

    /**
     * Sends a request to a target service and awaits its terminating response.
     *
     * @param baseUrl Base URL of the target service
     * @param build Builds the request message from the generated request id
     * @param onStream Invoked for each intermediate message, such as the per-file messages
     *        of a bulk load, before the terminating response arrives
     * @return The successful response
     * @throws ExecutionWsException if the peer answered with an error or could not be reached
     */
    suspend fun request(
        baseUrl: String,
        build: (requestId: String) -> ExecutionWsMessage,
        onStream: (suspend (ExecutionWsMessage) -> Unit)? = null
    ): ExecutionWsResponse {
        val url = toWebSocketUrl(baseUrl)
        // A cached connection can be closed by the peer between the moment it is handed out
        // and the moment it is written to, which is indistinguishable from it never having
        // worked. Retrying once on a fresh connection separates the two.
        return try {
            sendOnce(url, build, onStream)
        } catch (e: ExecutionWsException) {
            throw e
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            logger.warn("Execution WS request to {} failed ({}), retrying on a new connection", url, e.toString())
            connections.remove(url)?.close()
            try {
                sendOnce(url, build, onStream)
            } catch (retry: ExecutionWsException) {
                throw retry
            } catch (retry: CancellationException) {
                throw retry
            } catch (retry: Exception) {
                throw ExecutionWsException(
                    ExecutionWsErrorCodes.UNAVAILABLE,
                    "Execution WebSocket request to $url failed: ${retry.message}",
                    retry
                )
            }
        }
    }

    private suspend fun sendOnce(
        url: String,
        build: (requestId: String) -> ExecutionWsMessage,
        onStream: (suspend (ExecutionWsMessage) -> Unit)?
    ): ExecutionWsResponse {
        val connection = obtainConnection(url)
        val requestId = UUID.randomUUID().toString()
        val pending = PendingRequest(onStream)
        connection.pending[requestId] = pending
        try {
            connection.send(ExecutionWsProtocol.encode(build(requestId)))
            return withTimeout(requestTimeoutMillis) { pending.result.await() }
        } finally {
            connection.pending.remove(requestId)
        }
    }

    private suspend fun obtainConnection(url: String): PooledConnection {
        connections[url]?.let { if (it.isAlive) return it.touch() }
        return connectMutex.withLock {
            connections[url]?.let { if (it.isAlive) return@withLock it.touch() }
            connections.remove(url)?.close()
            val session = try {
                client.webSocketSession(url)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                logger.warn("Failed to open execution WS connection to {}: {}", url, e.toString())
                throw e
            }
            val connection = PooledConnection(url, session)
            connection.startReadLoop()
            connections[url] = connection
            connection
        }
    }

    private fun reapIdleConnections() {
        val now = System.currentTimeMillis()
        for ((url, connection) in connections) {
            val idle = now - connection.lastUsedAt
            if (!connection.isAlive) {
                connections.remove(url)?.close()
            } else if (idle >= idleTimeoutMillis && connection.pending.isEmpty()) {
                logger.debug("Closing idle execution WS connection to {} after {}ms", url, idle)
                connections.remove(url)?.close()
            }
        }
    }

    /**
     * Closes every pooled connection and, if this client created it, the underlying HTTP client.
     */
    override fun close() {
        reaper.cancel()
        for (url in connections.keys.toList()) {
            connections.remove(url)?.close()
        }
        scope.cancel()
        if (ownsHttpClient) {
            client.close()
        }
    }

    private inner class PooledConnection(
        private val url: String,
        private val session: DefaultClientWebSocketSession
    ) {
        val pending = ConcurrentHashMap<String, PendingRequest>()

        @Volatile
        var lastUsedAt: Long = System.currentTimeMillis()
            private set

        private var readLoop: Job? = null
        private val sendMutex = Mutex()

        val isAlive: Boolean
            get() = session.isActive && (readLoop?.isActive ?: true)

        fun touch(): PooledConnection {
            lastUsedAt = System.currentTimeMillis()
            return this
        }

        suspend fun send(text: String) {
            touch()
            sendMutex.withLock { session.send(Frame.Text(text)) }
        }

        fun startReadLoop() {
            readLoop = scope.launch {
                try {
                    for (frame in session.incoming) {
                        if (frame !is Frame.Text) continue
                        touch()
                        dispatch(ExecutionWsProtocol.decode(frame.readText()))
                    }
                    drain(ExecutionWsException(ExecutionWsErrorCodes.UNAVAILABLE, "Connection to $url closed"))
                } catch (e: CancellationException) {
                    drain(ExecutionWsException(ExecutionWsErrorCodes.UNAVAILABLE, "Connection to $url cancelled"))
                    throw e
                } catch (e: Exception) {
                    logger.warn("Execution WS read loop for {} ended: {}", url, e.toString())
                    drain(
                        ExecutionWsException(
                            ExecutionWsErrorCodes.UNAVAILABLE,
                            "Connection to $url failed: ${e.message}",
                            e
                        )
                    )
                } finally {
                    // Only evict this connection, never a replacement that was already
                    // opened for the same target after this one started failing.
                    if (connections[url] === this@PooledConnection) {
                        connections.remove(url)
                    }
                }
            }
        }

        private suspend fun dispatch(message: ExecutionWsMessage) {
            when (message) {
                is ExecutionWsResponse -> pending.remove(message.requestId)?.result?.complete(message)
                is ExecutionWsError -> pending.remove(message.requestId)?.result?.completeExceptionally(
                    ExecutionWsException(message.code, message.message)
                )
                else -> pending[message.requestId]?.onStream?.invoke(message)
            }
        }

        private fun drain(cause: Throwable) {
            for (key in pending.keys.toList()) {
                pending.remove(key)?.result?.completeExceptionally(cause)
            }
        }

        fun close() {
            readLoop?.cancel()
            drain(ExecutionWsException(ExecutionWsErrorCodes.UNAVAILABLE, "Connection to $url closed"))
            // Cancelled rather than closed politely: a close handshake would have to be
            // launched on this client's scope, which shutdown cancels out from under it,
            // leaving the socket open exactly when it most needs to go away.
            session.cancel()
        }
    }

    private class PendingRequest(val onStream: (suspend (ExecutionWsMessage) -> Unit)?) {
        val result = CompletableDeferred<ExecutionWsResponse>()
    }
}

/**
 * Failure of an execution WebSocket request, carrying the protocol error code so that the
 * originating failure keeps its meaning as it is relayed back up the chain of hops.
 *
 * @property code One of [ExecutionWsErrorCodes]
 */
class ExecutionWsException(
    val code: String,
    message: String,
    cause: Throwable? = null
) : RuntimeException(message, cause)

File: app/packages/service-common/src/ws/executionWsClient.ts
import { randomUUID } from "node:crypto";
import { COMPRESSION_THRESHOLD_BYTES } from "../util/compression.js";
import { WebSocket } from "ws";
import {
    ExecutionWsErrorCodes,
    ExecutionWsRequestError,
    toExecutionWsUrl,
    type ExecutionWsMessage,
    type ExecutionWsRequest,
    type ExecutionWsResponse
} from "./protocol.js";

/**
 * Default idle lifetime of a pooled connection. Long enough that the bursts of requests one
 * user interaction produces all share a connection, short enough that an idle service is not
 * holding sockets open for sessions nobody is using.
 */
const DEFAULT_IDLE_TIMEOUT_MS = 180_000;

/**
 * Default per-request timeout. Bulk loads of a large result set are the slow case this has
 * to accommodate.
 */
const DEFAULT_REQUEST_TIMEOUT_MS = 300_000;

/**
 * How often the reaper looks for connections that have gone idle.
 */
const REAPER_INTERVAL_MS = 30_000;

/**
 * How long to wait for a connection to be established.
 */
const CONNECT_TIMEOUT_MS = 10_000;

/**
 * Largest message accepted on the execution endpoint. Result files are sent whole, and a
 * large model can be tens of megabytes.
 */
const MAX_PAYLOAD_BYTES = 512 * 1024 * 1024;

/**
 * Reports connection failures.
 *
 * These connections are opened from module scope by the handlers that proxy to an execution
 * service, long before any Fastify request context exists, so this writes to the console
 * rather than the service logger.
 */
const log = {
    warn: (message: string): void => {
        // eslint-disable-next-line no-console
        console.warn(`[execution-ws] ${message}`);
    }
};

/**
 * Configuration for an {@link ExecutionWsClient}.
 */
export interface ExecutionWsClientOptions {
    /**
     * How long a connection may sit without traffic before it is closed.
     */
    idleTimeoutMs?: number;
    /**
     * How long a single request may wait for its terminating response.
     */
    requestTimeoutMs?: number;
}

/**
 * A request awaiting its terminating response.
 */
interface PendingRequest {
    resolve: (response: ExecutionWsResponse) => void;
    reject: (error: Error) => void;
    onStream?: (message: ExecutionWsMessage) => void;
    timer: NodeJS.Timeout;
}

/**
 * A connection held open for reuse between requests to one target service.
 */
class PooledConnection {
    readonly pending = new Map<string, PendingRequest>();
    lastUsedAt = Date.now();
    private socket: WebSocket | undefined;
    private opening: Promise<WebSocket> | undefined;

    constructor(private readonly url: string) {}

    /**
     * Returns the open socket, dialling if this connection has none yet.
     *
     * @returns The open socket
     */
    async open(): Promise<WebSocket> {
        if (this.socket && this.socket.readyState === WebSocket.OPEN) {
            this.lastUsedAt = Date.now();
            return this.socket;
        }
        if (this.opening) {
            return this.opening;
        }

        this.opening = new Promise<WebSocket>((resolve, reject) => {
            const socket = new WebSocket(this.url, {
                maxPayload: MAX_PAYLOAD_BYTES,
                perMessageDeflate: { threshold: COMPRESSION_THRESHOLD_BYTES }
            });
            const timer = setTimeout(() => {
                log.warn(`timed out connecting to ${this.url} after ${CONNECT_TIMEOUT_MS}ms`);
                socket.terminate();
                reject(new Error(`Timed out connecting to ${this.url}`));
            }, CONNECT_TIMEOUT_MS);

            socket.on("open", () => {
                clearTimeout(timer);
                this.socket = socket;
                this.lastUsedAt = Date.now();
                resolve(socket);
            });
            socket.on("message", (raw) => this.handleMessage(String(raw)));
            socket.on("error", (error) => {
                clearTimeout(timer);
                log.warn(`connection to ${this.url} errored: ${String(error)}`);
                reject(error);
            });
            socket.on("close", () => {
                clearTimeout(timer);
                this.socket = undefined;
                this.opening = undefined;
                if (this.pending.size > 0) {
                    log.warn(`connection to ${this.url} closed with ${this.pending.size} request(s) in flight`);
                }
                this.drain(
                    new ExecutionWsRequestError(ExecutionWsErrorCodes.Unavailable, `Connection to ${this.url} closed`)
                );
            });
        }).finally(() => {
            this.opening = undefined;
        });

        return this.opening;
    }

    /**
     * Reports whether this connection is usable without redialling.
     *
     * @returns true if the socket is open
     */
    get isOpen(): boolean {
        return this.socket?.readyState === WebSocket.OPEN;
    }

    /**
     * Sends a message on this connection.
     *
     * @param message The message to send
     */
    async send(message: ExecutionWsRequest): Promise<void> {
        const socket = await this.open();
        this.lastUsedAt = Date.now();
        socket.send(JSON.stringify(message));
    }

    /**
     * Closes the socket and fails everything still waiting on it.
     */
    close(): void {
        this.drain(new ExecutionWsRequestError(ExecutionWsErrorCodes.Unavailable, `Connection to ${this.url} closed`));
        this.socket?.close();
        this.socket = undefined;
    }

    private handleMessage(raw: string): void {
        this.lastUsedAt = Date.now();

        let message: ExecutionWsMessage;
        try {
            message = JSON.parse(raw) as ExecutionWsMessage;
        } catch {
            return;
        }

        const pending = this.pending.get(message.requestId);
        if (!pending) {
            return;
        }

        if (message.messageType === "exec/response") {
            this.settle(message.requestId, () => pending.resolve(message));
        } else if (message.messageType === "exec/error") {
            this.settle(message.requestId, () =>
                pending.reject(new ExecutionWsRequestError(message.code, message.message))
            );
        } else {
            pending.onStream?.(message);
        }
    }

    private settle(requestId: string, action: () => void): void {
        const pending = this.pending.get(requestId);
        if (!pending) {
            return;
        }
        clearTimeout(pending.timer);
        this.pending.delete(requestId);
        action();
    }

    private drain(error: Error): void {
        for (const requestId of [...this.pending.keys()]) {
            const pending = this.pending.get(requestId);
            if (pending) {
                clearTimeout(pending.timer);
                this.pending.delete(requestId);
                pending.reject(error);
            }
        }
    }
}

/**
 * Request/response client for the execution protocol, over WebSocket connections that are
 * cached per target service and closed again once they go unused.
 *
 * A connection is opened on the first request to a target, shared by every request to that
 * target afterwards, and closed after the configured idle timeout. Nothing about a request
 * depends on connection state — each one carries its own authorization — so a connection
 * that was dropped in between is simply reopened, and two unrelated callers can share one
 * without either being able to observe the other.
 */
export class ExecutionWsClient {
    private readonly connections = new Map<string, PooledConnection>();
    private readonly idleTimeoutMs: number;
    private readonly requestTimeoutMs: number;
    private readonly reaper: NodeJS.Timeout;

    /**
     * Creates a new client.
     *
     * @param options Idle and request timeouts; sensible defaults are used when omitted
     */
    constructor(options: ExecutionWsClientOptions = {}) {
        this.idleTimeoutMs = options.idleTimeoutMs ?? DEFAULT_IDLE_TIMEOUT_MS;
        this.requestTimeoutMs = options.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS;
        this.reaper = setInterval(() => this.reapIdleConnections(), REAPER_INTERVAL_MS);
        // The reaper must not be what keeps the process alive.
        this.reaper.unref?.();
    }

    /**
     * Sends a request to a target service and awaits its terminating response.
     *
     * @param baseUrl Base URL of the target service
     * @param build Builds the request message from the generated request id
     * @param onStream Invoked for each intermediate message, such as the per-file messages
     *        of a bulk load, before the terminating response arrives
     * @returns The successful response
     * @throws ExecutionWsRequestError if the peer answered with an error or could not be reached
     */
    async request(
        baseUrl: string,
        build: (requestId: string) => ExecutionWsRequest,
        onStream?: (message: ExecutionWsMessage) => void
    ): Promise<ExecutionWsResponse> {
        const url = toExecutionWsUrl(baseUrl);
        // A cached connection can be closed by the peer between the moment it is handed out
        // and the moment it is written to, which is indistinguishable from it never having
        // worked. Retrying once on a fresh connection separates the two.
        try {
            return await this.sendOnce(url, build, onStream);
        } catch (error) {
            if (error instanceof ExecutionWsRequestError && error.code !== ExecutionWsErrorCodes.Unavailable) {
                throw error;
            }
            this.connections.get(url)?.close();
            this.connections.delete(url);
            try {
                return await this.sendOnce(url, build, onStream);
            } catch (retryError) {
                if (retryError instanceof ExecutionWsRequestError) {
                    throw retryError;
                }
                throw new ExecutionWsRequestError(
                    ExecutionWsErrorCodes.Unavailable,
                    `Execution WebSocket request to ${url} failed: ${String(retryError)}`
                );
            }
        }
    }

    /**
     * Closes every pooled connection and stops the idle reaper.
     */
    close(): void {
        clearInterval(this.reaper);
        for (const connection of this.connections.values()) {
            connection.close();
        }
        this.connections.clear();
    }

    private async sendOnce(
        url: string,
        build: (requestId: string) => ExecutionWsRequest,
        onStream?: (message: ExecutionWsMessage) => void
    ): Promise<ExecutionWsResponse> {
        let connection = this.connections.get(url);
        if (!connection) {
            connection = new PooledConnection(url);
            this.connections.set(url, connection);
        }

        const requestId = randomUUID();
        return new Promise<ExecutionWsResponse>((resolve, reject) => {
            const timer = setTimeout(() => {
                connection!.pending.delete(requestId);
                log.warn(`request ${requestId} to ${url} timed out after ${this.requestTimeoutMs}ms`);
                reject(
                    new ExecutionWsRequestError(
                        ExecutionWsErrorCodes.Unavailable,
                        `Execution WebSocket request ${requestId} to ${url} timed out`
                    )
                );
            }, this.requestTimeoutMs);

            connection!.pending.set(requestId, { resolve, reject, onStream, timer });

            connection!.send(build(requestId)).catch((error: unknown) => {
                clearTimeout(timer);
                connection!.pending.delete(requestId);
                log.warn(`failed to send request ${requestId} to ${url}: ${String(error)}`);
                reject(
                    new ExecutionWsRequestError(
                        ExecutionWsErrorCodes.Unavailable,
                        `Failed to send execution WebSocket request to ${url}: ${String(error)}`
                    )
                );
            });
        });
    }

    private reapIdleConnections(): void {
        const now = Date.now();
        for (const [url, connection] of [...this.connections.entries()]) {
            const idleFor = now - connection.lastUsedAt;
            if (connection.pending.size > 0) {
                continue;
            }
            if (!connection.isOpen || idleFor >= this.idleTimeoutMs) {
                connection.close();
                this.connections.delete(url);
            }
        }
    }
}
